src/libstore/build/outer-derivation-goal.cc

```diff
@@ -0,0 +1,158 @@
+#include "outer-derivation-goal.hh"
+#include "worker.hh"
+
+namespace nix {
+
+OuterDerivationGoal::OuterDerivationGoal(ref<SingleDerivedPath> drvReq,
+    const OutputsSpec & wantedOutputs, Worker & worker, BuildMode buildMode)
+    : Goal(worker, DerivedPath::Built { .drvPath = drvReq, .outputs = wantedOutputs })
+    , drvReq(drvReq)
+    , wantedOutputs(wantedOutputs)
+    , buildMode(buildMode)
+{
+    state = &OuterDerivationGoal::getDerivation;
+    name = fmt(
+        "outer obtaining drv from '%s' and then building outputs %s",
+        drvReq->to_string(worker.store),
+        std::visit(overloaded {
+            [&](const OutputsSpec::All) -> std::string {
+                return "* (all of them)";
+            },
+            [&](const OutputsSpec::Names os) {
+                return concatStringsSep(", ", quoteStrings(os));
+            },
+        }, wantedOutputs.raw()));
+    trace("created outer");
+
+    worker.updateProgress();
+}
+
+
+OuterDerivationGoal::~OuterDerivationGoal()
+{
+}
+
+
+static StorePath pathPartOfReq(const SingleDerivedPath & req)
+{
+    return std::visit(overloaded {
+        [&](const SingleDerivedPath::Opaque & bo) {
+            return bo.path;
+        },
+        [&](const SingleDerivedPath::Built & bfd) {
+            return pathPartOfReq(*bfd.drvPath);
+        },
+    }, req.raw());
+}
+
+
+std::string OuterDerivationGoal::key()
+{
+    /* Ensure that derivations get built in order of their name,
```

Comment:
Why is this needed? Seems a strange thing to depend on the name.

Reply:
The other goals already have it, I just immitated it. It would seem to be messing with the order for tree maps, so iteration in order gets some types of jobs before other types of jobs?

Reply:
Perhaps now that we have `jobCategory()` we can use that in `GoalPtr` comparison instead? But only if we make every type of goal have its own category.

Reply:
> only if we make every type of goal have its own category.

That would make `jobCategory()` hard to use for its current purpose of having separate limits for substitution and building.

A separate method for this purpose could be ok, but I don't know enough about "`GoalPtr` comparison" to weigh it.

Reply:
That is true, but also right now we are effectively double-counting every derivation, since there is both a `CreateDerivationAndRealiseGoal` and `DerivationGoal` for each one. So resolving that would have us have a 3rd category anyways?

Reply:
(Converted to draft since I realized this issue.)

Reply:
That issue is now fixed.